requirements-training.txt

```diff
@@ -4,6 +4,6 @@ h5py
 numpy >= 1.16.6
 onnx
 packaging
-protobuf >= 3.12.2, <= 3.20.1
+protobuf >= 3.20.0, <= 3.21.0
```

Comment:
That doesn't looks right. <= 3.21.0 means you want 3.21.0 (although it won't exist because they call it 4.21.0) but not 3.21.1

Reply:
thanks for sharing this. I'll make it < 4 in that case.

Reply:
Better to be <4, because that clearly states "major version introduces breaking change"

Reply:
And no it's not <4.21.1, 4.21.0 won't work either.
What happen is, protobuf had a breaking change in 3.21, then its python variant decided to name 3.21 as 4.21 in pip wheel, and there're still new releases of 3.20.x after 4.21.